### PR TITLE
Added section entropy & flags + option to log import table

### PR DESCRIPTION
**Pull request by fox-ds**
This PR adds the following features:

- Process the section bytes of all sections (with a limit of 5MiB per section) and produce a count of byte occurrences.
- Enable the option of calculating & logging the section Shannon entropy ( so not just entropy over the entire file, which was already possible in `files.log` ).
- Enable the option of logging of the section flags (read/write/execute) for every section.
- Enable the option of logging the entire import table per PE.

By default, this extra logging is disable. By overriding some values, you can enable the extra logs.
```zeek
export {
    redef PE::pe_log_section_entropy=T;
    redef PE::pe_log_section_flags=T;
    redef PE::pe_log_import_table=T;
}
```

which will e.g. produce the following result:
```json
{ "section_names": [                                                                                                                                                                                                                                                                                  
    ".text:er:6.03",                                                                                                                                                  
    ".data:wr:7.14",                                                                                                                                                  
    ".pdata:r:2.97",                                                                                                                                                  
    ".idata:wr:4.14",                                                                                                                                                 
    ".CRT:wr:0.26",                                                                                                                                                                                                                                                                                        
    ".rdata:r:4.00",                                                                                                                                                                                                                                                                    
  ],                                                                                                                                                                  
  "import_table": [                                                                                                                                                   
    "KERNEL32.dll:CloseHandle",                                                                                                                                                                                                                                                                                                                                                                                      
    "KERNEL32.dll:CreateNamedPipeA"
  ],
}
```

Because Zeek is not able to log nested maps, the solution of appending all info (name+flags+entropy) was chosen over logging hem individually, for efficiency and data-duplication reasons. For convenience, we just override the `section_names` that was already present and extend that with extra information.

The calculation of entropy was split between Spicy/Zeek because Spicy currently does not have a function to calculate logarithms. That why some of the heavy lifting (counting of byte occurences) is done in Spicy, before those counts are passed to Zeek which does the more complex calculations.

### CODE REVIEW

**Comment by fox-ds** [comment]
I saw that the tests raised an error, but the CI passed.

`error: "/__w/spicy-pe/spicy-pe" tests failed, inspect contents of /opt/zeek/var/lib/zkg/testing/spicy-pe for details, especially any "zkg.test_command.{stderr,stdout}" files within /opt/zeek/var/lib/zkg/testing/spicy-pe/clones/spicy-pe`

**Comment by fox-ds** [comment]
Right, the unit tests indeed fail. I'll fix them.

File: analyzer/main.zeek [hunk 1]
```diff
@@ -50,9 +65,137 @@ export {
...
+			if (pe_log_section_entropy && ! pe_log_section_flags ) {
+				formatted_string = fmt("%s:%.2f", section, info$entropy);
+			}
+			if (! pe_log_section_entropy && pe_log_section_flags ) {
+				formatted_string = fmt("%s:%s", section, info$flags);
+			}
+			f$pe$section_names += formatted_string;
+		}
+	}
+}
```
Comment (bbannier) [inline analyzer/main.zeek:196]: The different conditionals inside the loop are getting hard to follow, how about the following instead?
```suggestion

# Called when the file analysis is closed
event file_state_remove(f: fa_file)
    {
	# If any of the detailed logging is enabled, delete the default section_names field.
	# This means that default functionality is not changed.
	if ( pe_log_section_flags  || pe_log_section_entropy )
		{
		f$pe$section_names = vector();

		for ( section, info in f$pe$section_info_table )
			{
			local log_entropy = pe_log_section_entropy ? fmt("%.2f", info$entropy) : "";
			local log_section_flags = pe_log_section_flags ? fmt("%s", info$flags) : "";
			local log_frag = vector(log_section_flags, log_entropy);

			local formatted_string: string = fmt("%s", section);
			for (i in log_frag)
				{
				if (|log_frag[i]| != 0)
					{
					formatted_string += fmt(":%s", log_frag[i]);
					}
				}

			f$pe$section_names += formatted_string;
			}
		}
	}
```

File: analyzer/main.zeek [hunk 2]
```diff
@@ -50,9 +65,137 @@ export {
...
 	};
+}
+
+function shannon_entropy(counts: table[count] of count, sectionTotalBytes: double) : double {
+	local entropy: double = 0.0;
+
+	# Calculate the Shannon entropy of the bits
+	# https://en.wikipedia.org/wiki/Entropy_(information_theory)
+	# H(X) = -sum(P_xi * log_2(xi))
+	# where log2() is represented with log10(p_x)/log10(2)
```
Comment (bbannier) [inline analyzer/main.zeek:76]:
```suggestion
	# Calculate the Shannon entropy of the bits,
	# https://en.wikipedia.org/wiki/Entropy_(information_theory):
	#
	#   H(X) = -sum(P_xi * log_2(P_xi))
	#
	# where log2(p_x) is represented with log10(p_x)/log10(2).
```

File: analyzer/main.zeek [hunk 3]
```diff
@@ -50,9 +65,137 @@ export {
...
+				formatted_string = fmt("%s:%.2f", section, info$entropy);
+			}
+			if (! pe_log_section_entropy && pe_log_section_flags ) {
+				formatted_string = fmt("%s:%s", section, info$flags);
+			}
+			f$pe$section_names += formatted_string;
+		}
+	}
+}
+
```
Comment (bbannier) [inline analyzer/main.zeek:197]:
```suggestion
```

**Review by bbannier — CHANGES_REQUESTED** [review]
Thanks for this PR. This looks great.

I was wondering whether this would be useful enough for general consumption to include at least the Spicy bits on the default parsing path, but sadly we do not have a solution or zeek/spicy#765 yet, but the extra work seems to be acceptable.

I left only minor cleanup comments below.

File: analyzer/analyzer.spicy [hunk 4]
```diff
@@ -134,6 +183,8 @@ public type ImageFile = unit {
     var sectionDataSize: uint64;
     var exportDirInfo: DirInfo;
     var importDirInfo: DirInfo;
+    var sectionCounts: map<string, map<uint8, uint64>>;
+    var sectionLengths: map<string, real>;
```
[the remark this answers is not in the document]
Comment (fox-ds) [reply analyzer/analyzer.spicy]: Yes, I added some comments to the variables.

File: analyzer/analyzer.spicy [hunk 5]
```diff
@@ -126,6 +126,55 @@ public type ImageFile = unit {
             }
         }
 
+    on sectionData
+        {
+            # Initialize the map that will hold the counts
+            self.sectionCounts =  map<string, map<uint8, uint64>>();
+            self.sectionLengths =  map<string, real>();
+
```
[the remark this answers is not in the document]
Comment (fox-ds) [reply analyzer/analyzer.spicy]: Removed the lines

File: analyzer/analyzer.spicy [hunk 6]
```diff
@@ -126,6 +126,55 @@ public type ImageFile = unit {
...
+            for ( sh in self.sectionHeaders )
+            {
+              local section_name = sh.name.strip(spicy::Side::Right, b"\x00").decode();
+
+              if ( sh.pointerToRawData != 0 )
+                {
+                # Get the correct raw bytes of this section by calculating the location
+                # of the bytes relative to the current offset
+                local relativeStart: uint64 = sh.pointerToRawData - off;
```
[the remark this answers is not in the document]
Comment (fox-ds) [reply analyzer/analyzer.spicy:151]: Added an extra check here.

File: analyzer/analyzer.spicy [hunk 7]
```diff
@@ -126,6 +126,55 @@ public type ImageFile = unit {
...
+
+                local nBytesToProcess: uint32 = 0;
+
+                # Cut off after 5MiB, since by then you probably have a good idea of whether the section
+                # has high or low entropy. This is done to increase performance. In the case of a large section,
+                # are then calculating the entropy of the first 5MiB of that section.
+                if ( sh.sizeOfRawData > 5242880 ) {
```
[the remark this answers is not in the document]
Comment (fox-ds) [reply analyzer/analyzer.spicy]: Agree, added a constant for this.

File: analyzer/analyzer.spicy [hunk 8]
```diff
@@ -126,6 +126,55 @@ public type ImageFile = unit {
...
+                    nBytesToProcess = 5242880;
+                } else {
+                    nBytesToProcess = sh.sizeOfRawData;
+                }
+
+                local sectionBytes: bytes = $$.sub(relativeStart, relativeStart + nBytesToProcess);
+                local sectionCount: map<uint8, uint64> = calculateCounts(sectionBytes);
+                self.sectionCounts[section_name] = sectionCount;
+                self.sectionLengths[section_name] = cast<real>(|sectionBytes|);
```
[the remark this answers is not in the document]
Comment (fox-ds) [reply analyzer/analyzer.spicy]: Yes, that's perfectly possible. I changed it.

File: analyzer/main.zeek [hunk 9]
```diff
@@ -50,9 +65,137 @@ export {
...
+    }
+    # Finally, put it in the actual PE log
+    f$pe$import_table = temp_tbl;
 }
 
+# Called when the file analysis is closed
+event file_state_remove(f: fa_file)
+    {
+	# If any of the new logging is enabled, delete the default section_names field
+	# This means that default functionality is not changed
```
[the remark this answers is not in the document]
Comment (fox-ds) [reply analyzer/main.zeek]: Added punctation to the sentences and changed this line.

File: analyzer/main.zeek (hunk 1 above)
Comment (fox-ds) [reply analyzer/main.zeek]: Agree, I committed your suggestion.

File: analyzer/main.zeek (hunk 2 above)
Comment (fox-ds) [reply analyzer/main.zeek]: Done

**Pull request merged**